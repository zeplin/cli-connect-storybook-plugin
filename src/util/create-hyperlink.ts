import urlJoin from "proper-url-join";
import { toId } from "@storybook/csf";

interface ParamsWithStoryId {
    storyId: string;
    hasDocsPage?: boolean;
}

interface ParamsWithKindAndStory {
    selectedKind: string;
    selectedStory?: string;
}

interface StoryHyperlinkOptions {
    format: "old" | "new";
}

type StoryHyperlinkParams = ParamsWithStoryId | ParamsWithKindAndStory;

function isParamsWithKindAndStory(params: StoryHyperlinkParams): params is ParamsWithKindAndStory {
    return "selectedKind" in params;
}

function toStoryId(params: ParamsWithKindAndStory): string {
    const { selectedKind, selectedStory } = params;
    if (selectedStory) {
        return toId(selectedKind, selectedStory);
    }

    return toId(selectedKind, "placeholder").replace("placeholder", "*");
}

function toLegacyQuery(params: ParamsWithKindAndStory): { [key: string]: string } {
    const { selectedKind, selectedStory } = params;

    return selectedStory ? { selectedKind, selectedStory } : { selectedKind };
}

function createStoryHyperlink(
    baseUrl: string,
    params: StoryHyperlinkParams,
    options: StoryHyperlinkOptions = { format: "old" }
): string {
    let url: string;

    // Trailing slash is required if the Storybook is hosted under a path but if
    // BaseUrl just points to the iframe.html inside Storybook, trailing slash should not exist
    const trailingSlash = !baseUrl.endsWith("iframe.html");

    if (isParamsWithKindAndStory(params) && options.format === "new") {
        url = urlJoin(baseUrl, {
            trailingSlash,
            query: { path: `/story/${toStoryId(params)}` },
            queryOptions: { encode: false }
        });
    } else if (isParamsWithKindAndStory(params)) {
        url = urlJoin(baseUrl, {
            trailingSlash,
            query: toLegacyQuery(params)
        });
    } else {
        const { storyId } = params;

        // Docs hyperlinks somehow cause error if iframe is accessed directly
        // To workaround this /story/ is enforced even if a docs page exist
        const viewMode = params.hasDocsPage && !baseUrl.endsWith("iframe.html")
            ? "docs"
            : "story";

        url = urlJoin(baseUrl, {
            trailingSlash,
            query: { path: `/${viewMode}/${storyId}` },
            queryOptions: { encode: false }
        });
    }

    return url;
}

export {
    StoryHyperlinkParams,
    StoryHyperlinkOptions,
    createStoryHyperlink
};