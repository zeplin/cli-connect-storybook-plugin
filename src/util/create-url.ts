import urlJoin from "proper-url-join";

interface ParamsWithStoryId {
    storyId: string;
    hasDocsPage?: boolean;
}

interface ParamsWithKindAndStory {
    selectedKind: string;
    selectedStory?: string;
}

type StorybookLinkParams = ParamsWithStoryId | ParamsWithKindAndStory;

function isV5(params: StorybookLinkParams): params is ParamsWithStoryId {
    return "storyId" in params;
}

function createStorybookUrl(baseUrl: string, params: StorybookLinkParams): string {
    let url: string;

    if (isV5(params)) {
        const { storyId } = params;

        const viewMode = params.hasDocsPage ? "docs" : "story";

        url = urlJoin(baseUrl, {
            trailingSlash: true,
            query: { path: `/${viewMode}/${storyId}` },
            queryOptions: { encode: false }
        });
    } else {
        const { selectedKind, selectedStory } = params;

        if (selectedStory) {
            url = urlJoin(baseUrl, {
                trailingSlash: true,
                query: { selectedKind, selectedStory }
            });
        } else {
            url = urlJoin(baseUrl, {
                trailingSlash: true,
                query: { selectedKind }
            });
        }
    }

    return url;
}

export {
    StorybookLinkParams,
    createStorybookUrl
};