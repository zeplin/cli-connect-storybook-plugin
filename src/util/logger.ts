// eslint-disable-next-line import/named
import { Logger } from "@zeplin/cli";

let loggerInstance: Logger = {
    error: console.error,
    info: console.info,
    warn: console.warn,
    debug: console.debug
};

const getLogger = (): Logger => loggerInstance;
const setLogger = (logger?: Logger): void => {
    if (logger) {
        loggerInstance = logger;
    }
};

export { getLogger, setLogger };
