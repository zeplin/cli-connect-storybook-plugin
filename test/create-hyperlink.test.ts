import { createStoryHyperlink } from "../src/util/create-hyperlink";

const BASE_URL = "http://localhost:9009";
const BASE_URL_WITH_TRAILING_SLASH = "http://localhost:9009/";
const BASE_URL_PATH = "http://localhost:9009/hello";
const BASE_URL_PATH_WITH_TRAILING_SLASH = "http://localhost:9009/hello/";

const STORY_ID = "something-kind--story-hello-there";
const SELECTED_KIND = "something/kind";
const SELECTED_STORY = "story/hello*there!";

const EXPECTED_V5_HYPERLINK = "http://localhost:9009/?path=/story/something-kind--story-hello-there";
const EXPECTED_V5_DOCS_HYPERLINK = "http://localhost:9009/?path=/docs/something-kind--story-hello-there";
const EXPECTED_V5_HYPERLINK_ONLY_KIND = "http://localhost:9009/?path=/story/something-kind--*";
const EXPECTED_V4_HYPERLINK_KIND = "http://localhost:9009/?selectedKind=something%2Fkind";
const EXPECTED_V4_HYPERLINK_KIND_STORY = "http://localhost:9009/?selectedKind=something%2Fkind&selectedStory=story%2Fhello%2Athere%21";

const EXPECTED_V5_HYPERLINK_PATH = "http://localhost:9009/hello/?path=/story/something-kind--story-hello-there";
const EXPECTED_V5_DOCS_HYPERLINK_PATH = "http://localhost:9009/hello/?path=/docs/something-kind--story-hello-there";
const EXPECTED_V4_HYPERLINK_PATH_KIND = "http://localhost:9009/hello/?selectedKind=something%2Fkind";
const EXPECTED_V4_HYPERLINK_PATH_KIND_STORY = "http://localhost:9009/hello/?selectedKind=something%2Fkind&selectedStory=story%2Fhello%2Athere%21";

describe("createStoryHyperlink", () => {
    let baseUrl = BASE_URL;

    it("returns Storybook v5+ hyperlink if storyId is provided", () => {
        const hyperlink = createStoryHyperlink(baseUrl, { storyId: STORY_ID });

        expect(hyperlink).toBe(EXPECTED_V5_HYPERLINK);
    });

    it("returns Storybook v5+ docs hyperlink if storyId is provided and has docs page", () => {
        const hyperlink = createStoryHyperlink(baseUrl, { storyId: STORY_ID, hasDocsPage: true });

        expect(hyperlink).toBe(EXPECTED_V5_DOCS_HYPERLINK);
    });

    it("returns Storybook v5+ hyperlink if both selectedKind and selectedStory is provided", () => {
        const hyperlink = createStoryHyperlink(
            baseUrl,
            { selectedKind: SELECTED_KIND, selectedStory: SELECTED_STORY },
            { format: "new" }
        );

        expect(hyperlink).toBe(EXPECTED_V5_HYPERLINK);
    });

    it("returns Storybook v5+ hyperlink if only selectedKind is provided", () => {
        const hyperlink = createStoryHyperlink(
            baseUrl,
            { selectedKind: SELECTED_KIND },
            { format: "new" }
        );

        expect(hyperlink).toBe(EXPECTED_V5_HYPERLINK_ONLY_KIND);
    });

    it("returns Storybook v4 hyperlink if only selectedKind is provided", () => {
        const hyperlink = createStoryHyperlink(baseUrl, { selectedKind: SELECTED_KIND });

        expect(hyperlink).toBe(EXPECTED_V4_HYPERLINK_KIND);
    });

    it("returns Storybook v4 hyperlink if both selectedKind and selectedStory is provided", () => {
        const hyperlink = createStoryHyperlink(baseUrl, {
            selectedKind: SELECTED_KIND,
            selectedStory: SELECTED_STORY
        });

        expect(hyperlink).toBe(EXPECTED_V4_HYPERLINK_KIND_STORY);
    });

    describe("when baseUrl has trailing slash", () => {
        beforeAll(() => {
            baseUrl = BASE_URL_WITH_TRAILING_SLASH;
        });

        it("returns Storybook v5+ hyperlink if storyId is provided", () => {
            const hyperlink = createStoryHyperlink(baseUrl, { storyId: STORY_ID });

            expect(hyperlink).toBe(EXPECTED_V5_HYPERLINK);
        });

        it("returns Storybook v5+ docs hyperlink if storyId is provided and has docs page", () => {
            const hyperlink = createStoryHyperlink(baseUrl, { storyId: STORY_ID, hasDocsPage: true });

            expect(hyperlink).toBe(EXPECTED_V5_DOCS_HYPERLINK);
        });

        it("returns Storybook v4 hyperlink if only selectedKind is provided", () => {
            const hyperlink = createStoryHyperlink(baseUrl, { selectedKind: SELECTED_KIND });

            expect(hyperlink).toBe(EXPECTED_V4_HYPERLINK_KIND);
        });

        it("returns Storybook v4 hyperlink if both selectedKind and selectedStory is provided", () => {
            const hyperlink = createStoryHyperlink(baseUrl, {
                selectedKind: SELECTED_KIND,
                selectedStory: SELECTED_STORY
            });

            expect(hyperlink).toBe(EXPECTED_V4_HYPERLINK_KIND_STORY);
        });
    });

    describe("when baseUrl has path", () => {
        beforeAll(() => {
            baseUrl = BASE_URL_PATH;
        });

        it("returns Storybook v5+ hyperlink if storyId is provided", () => {
            const hyperlink = createStoryHyperlink(baseUrl, { storyId: STORY_ID });

            expect(hyperlink).toBe(EXPECTED_V5_HYPERLINK_PATH);
        });

        it("returns Storybook v5+ docs hyperlink if storyId is provided and has docs page", () => {
            const hyperlink = createStoryHyperlink(baseUrl, { storyId: STORY_ID, hasDocsPage: true });

            expect(hyperlink).toBe(EXPECTED_V5_DOCS_HYPERLINK_PATH);
        });

        it("returns Storybook v4 hyperlink if only selectedKind is provided", () => {
            const hyperlink = createStoryHyperlink(baseUrl, { selectedKind: SELECTED_KIND });

            expect(hyperlink).toBe(EXPECTED_V4_HYPERLINK_PATH_KIND);
        });

        it("returns Storybook v4 hyperlink if both selectedKind and selectedStory is provided", () => {
            const hyperlink = createStoryHyperlink(baseUrl, {
                selectedKind: SELECTED_KIND,
                selectedStory: SELECTED_STORY
            });

            expect(hyperlink).toBe(EXPECTED_V4_HYPERLINK_PATH_KIND_STORY);
        });
    });

    describe("when baseUrl has path with trailing slash", () => {
        beforeAll(() => {
            baseUrl = BASE_URL_PATH_WITH_TRAILING_SLASH;
        });

        it("returns Storybook v5+ hyperlink if storyId is provided", () => {
            const hyperlink = createStoryHyperlink(baseUrl, { storyId: STORY_ID });

            expect(hyperlink).toBe(EXPECTED_V5_HYPERLINK_PATH);
        });

        it("returns Storybook v5+ docs hyperlink if storyId is provided and has docs page", () => {
            const hyperlink = createStoryHyperlink(baseUrl, { storyId: STORY_ID, hasDocsPage: true });

            expect(hyperlink).toBe(EXPECTED_V5_DOCS_HYPERLINK_PATH);
        });

        it("returns Storybook v4 hyperlink if only selectedKind is provided", () => {
            const hyperlink = createStoryHyperlink(baseUrl, { selectedKind: SELECTED_KIND });

            expect(hyperlink).toBe(EXPECTED_V4_HYPERLINK_PATH_KIND);
        });

        it("returns Storybook v4 hyperlink if both selectedKind and selectedStory is provided", () => {
            const hyperlink = createStoryHyperlink(baseUrl, {
                selectedKind: SELECTED_KIND,
                selectedStory: SELECTED_STORY
            });

            expect(hyperlink).toBe(EXPECTED_V4_HYPERLINK_PATH_KIND_STORY);
        });
    });
});