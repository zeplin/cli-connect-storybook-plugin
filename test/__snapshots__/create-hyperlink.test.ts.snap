// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStoryHyperlink with base url http://localhost:9009 returns Storybook v4 hyperlink if format is old 1`] = `"http://localhost:9009/?selectedKind=something%2Fkind&selectedStory=story%2Fhello%2Athere%21"`;

exports[`createStoryHyperlink with base url http://localhost:9009 returns Storybook v5+ docs hyperlink if storyId is provided and has docs page 1`] = `"http://localhost:9009/?path=/docs/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009 returns Storybook v5+ docs hyperlink if storyId is provided and has docs page but useDocsPage is falsy 1`] = `"http://localhost:9009/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009 returns Storybook v5+ docs hyperlink if storyId is provided and has docs page but useDocsPage is falsy 2`] = `"http://localhost:9009/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009 returns Storybook v5+ hyperlink if storyId is provided 1`] = `"http://localhost:9009/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/ returns Storybook v4 hyperlink if format is old 1`] = `"http://localhost:9009/?selectedKind=something%2Fkind&selectedStory=story%2Fhello%2Athere%21"`;

exports[`createStoryHyperlink with base url http://localhost:9009/ returns Storybook v5+ docs hyperlink if storyId is provided and has docs page 1`] = `"http://localhost:9009/?path=/docs/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/ returns Storybook v5+ docs hyperlink if storyId is provided and has docs page but useDocsPage is falsy 1`] = `"http://localhost:9009/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/ returns Storybook v5+ docs hyperlink if storyId is provided and has docs page but useDocsPage is falsy 2`] = `"http://localhost:9009/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/ returns Storybook v5+ hyperlink if storyId is provided 1`] = `"http://localhost:9009/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/hello returns Storybook v4 hyperlink if format is old 1`] = `"http://localhost:9009/hello/?selectedKind=something%2Fkind&selectedStory=story%2Fhello%2Athere%21"`;

exports[`createStoryHyperlink with base url http://localhost:9009/hello returns Storybook v5+ docs hyperlink if storyId is provided and has docs page 1`] = `"http://localhost:9009/hello/?path=/docs/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/hello returns Storybook v5+ docs hyperlink if storyId is provided and has docs page but useDocsPage is falsy 1`] = `"http://localhost:9009/hello/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/hello returns Storybook v5+ docs hyperlink if storyId is provided and has docs page but useDocsPage is falsy 2`] = `"http://localhost:9009/hello/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/hello returns Storybook v5+ hyperlink if storyId is provided 1`] = `"http://localhost:9009/hello/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/hello/ returns Storybook v4 hyperlink if format is old 1`] = `"http://localhost:9009/hello/?selectedKind=something%2Fkind&selectedStory=story%2Fhello%2Athere%21"`;

exports[`createStoryHyperlink with base url http://localhost:9009/hello/ returns Storybook v5+ docs hyperlink if storyId is provided and has docs page 1`] = `"http://localhost:9009/hello/?path=/docs/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/hello/ returns Storybook v5+ docs hyperlink if storyId is provided and has docs page but useDocsPage is falsy 1`] = `"http://localhost:9009/hello/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/hello/ returns Storybook v5+ docs hyperlink if storyId is provided and has docs page but useDocsPage is falsy 2`] = `"http://localhost:9009/hello/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/hello/ returns Storybook v5+ hyperlink if storyId is provided 1`] = `"http://localhost:9009/hello/?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/iframe.html returns Storybook v4 hyperlink if format is old 1`] = `"http://localhost:9009/iframe.html?selectedKind=something%2Fkind&selectedStory=story%2Fhello%2Athere%21"`;

exports[`createStoryHyperlink with base url http://localhost:9009/iframe.html returns Storybook v5+ docs hyperlink if storyId is provided and has docs page 1`] = `"http://localhost:9009/iframe.html?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/iframe.html returns Storybook v5+ docs hyperlink if storyId is provided and has docs page but useDocsPage is falsy 1`] = `"http://localhost:9009/iframe.html?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/iframe.html returns Storybook v5+ docs hyperlink if storyId is provided and has docs page but useDocsPage is falsy 2`] = `"http://localhost:9009/iframe.html?path=/story/something-kind--story-hello-there"`;

exports[`createStoryHyperlink with base url http://localhost:9009/iframe.html returns Storybook v5+ hyperlink if storyId is provided 1`] = `"http://localhost:9009/iframe.html?path=/story/something-kind--story-hello-there"`;
